### PR TITLE
Fix many doc issues

### LINKED ISSUE #35643: nn.Module hooks ignore kwargs
To reproduce:

https://gist.github.com/zjplab/1d7b2a22a13d63f3ad83a59563cc4141

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345 @saketh-are @ezyang @SsnL @zou3519 @gqchen

### PR DESCRIPTION
**Pull request by albanD**
Fix https://github.com/pytorch/pytorch/issues/35643 https://github.com/pytorch/pytorch/issues/37063 https://github.com/pytorch/pytorch/issues/36307 https://github.com/pytorch/pytorch/issues/35861 https://github.com/pytorch/pytorch/issues/35299 https://github.com/pytorch/pytorch/issues/23108 https://github.com/pytorch/pytorch/issues/4661

Just a bunch of small updates on the doc.

### CODE REVIEW

File: docs/source/notes/extending.rst [hunk 1]
```diff
@@ -121,6 +121,10 @@ non-Tensor arguments::
     track history. So if ``backward`` is implemented with differentiable
     operations, (e.g., invocation of another custom
     :class:`~torch.autograd.function`), higher order derivatives will work.
+    In this case, the Tensors saved with ``save_for_backward`` can also be used
+    in a differentiable manner but Tensor saved in the ``ctx`` cannot be used.
```
Comment (mruberry) [inline docs/source/notes/extending.rst:125]: "in a differentiable manner" - maybe clarify this? Can you be more specific about what tensors saved in 'ctx' cannot be used for? They cannot be part of a backward graph, right?

File: docs/source/notes/extending.rst [hunk 2]
```diff
@@ -121,6 +121,10 @@ non-Tensor arguments::
...
+    If you need one, you should make it an output of the custom ``Function`` and
```
Comment (mruberry) [inline docs/source/notes/extending.rst:126]: one what? One of the tensors saved in ctx?

File: torch/autograd/__init__.py [hunk 3]
```diff
@@ -57,6 +57,13 @@ def backward(tensors, grad_tensors=None, retain_graph=None, create_graph=False,
     This function accumulates gradients in the leaves - you might need to zero
     them before calling it.
 
+    .. note::
+        Using this method with ``create_graph=True`` will create a reference cycle
+        between the parameter and its gradient which can lead to memory leak.
```
Comment (mruberry) [inline torch/autograd/__init__.py:62]: "can cause a memory leak" (you're missing an 'a')

File: torch/autograd/__init__.py [hunk 4]
```diff
@@ -57,6 +57,13 @@ def backward(tensors, grad_tensors=None, retain_graph=None, create_graph=False,
...
+        We recommend using ``autograd.grad`` below when creating the graph to avoid this.
```
Comment (mruberry) [inline torch/autograd/__init__.py:63]: "below" seems oddly placed in this sentence. Aren't you recommending it to me here?

File: torch/functional.py [hunk 5]
```diff
@@ -237,6 +237,14 @@ def einsum(equation, *operands):
            the ellipsis dimensions are at the beginning of the output.
     operands (Tensor): The operands to compute the Einstein sum of.
 
+.. note::
+
+    This function does not perform any memory or speed optimization on the given expression.
+    So doing the same overall computation with a different formula can lead to different memory
```
Comment (mruberry) [inline torch/functional.py:243]: This function does not optimize the given expression, so a different formula for the same computation may run faster or consume less memory. Projects like opt_einsum (https://optimized-einsum.readthedocs.io/en/stable/) can optimize the formula for you.

File: torch/nn/modules/module.py [hunk 6]
```diff
@@ -462,7 +462,8 @@ def register_backward_hook(self, hook):
         module has multiple inputs or outputs. The hook should not modify its
         arguments, but it can optionally return a new gradient with respect to
         input that will be used in place of :attr:`grad_input` in subsequent
-        computations.
+        computations. :attr:`grad_input` will only correspond to the inputs given
```
Comment (mruberry) [inline torch/nn/modules/module.py:465]: I would make this a note, like: "Note: the grad_input tuple only..."

But also, can we be more specific about what grad_input does/doesn't contain? This note seems to refer to an implementation detail, and not to our user-facing contract?

File: torch/nn/modules/module.py [hunk 7]
```diff
@@ -491,6 +492,7 @@ def register_forward_pre_hook(self, hook):
 
             hook(module, input) -> None or modified input
 
+        The input contains only the positional arguments given to the module.
```
Comment (mruberry) [inline torch/nn/modules/module.py:495]: Can you get the kwargs? If not, maybe be explicit? What about a positional argument specified as a kwarg?

File: torch/nn/modules/module.py [hunk 8]
```diff
@@ -512,6 +514,7 @@ def register_forward_hook(self, hook):
 
             hook(module, input, output) -> None or modified output
 
+        The input contains only the positional arguments given to the module.
```
Comment (mruberry) [inline torch/nn/modules/module.py:517]: Same comment as above.

**Review by mruberry — APPROVED** [review]
Cool! Left some comments/suggestions that you can optionally take.

File: torch/nn/modules/module.py (hunk 6 above)
Comment (albanD) [reply torch/nn/modules/module.py:465]: I didn't want to elaborate to much here as these are broken anyway and grad_input usually contains random stuff :)

File: torch/nn/modules/module.py (hunk 7 above)
Comment (albanD) [reply torch/nn/modules/module.py:495]: You get everything captured by `*args` on the function call. Not sure if that contains positional arguments passed as kwargs?

File: torch/nn/modules/module.py (hunk 6 above)
Comment (mruberry) [reply torch/nn/modules/module.py:465]: Gasp!

File: torch/autograd/__init__.py (hunk 4 above)
Comment (albanD) [reply torch/autograd/__init__.py:63]: Yes, I meant the function in the doc just below. But it is unclear, removing it.

File: torch/nn/modules/module.py (hunk 6 above)
Comment (albanD) [reply torch/nn/modules/module.py:465]: This is not the function you're looking for !

File: torch/nn/modules/module.py (hunk 7 above)
Comment (albanD) [reply torch/nn/modules/module.py:495]: After testing, there is no nothing of positional as kwarg :D If you do `foo=2` when calling a function, it is a kwarg.

Comment (albanD) [reply torch/nn/modules/module.py:495]: Updating to explicitly state that the kwargs are ignored.

**Pull request closed without merging**